complete-firebase-authentication-react-tutorial.md

```diff
@@ -372,209 +360,385 @@ const config = {
+The Firebase Context from the Firebase module (folder) is used to provide a Firebase instance to your entire application in the *src/index.js* file. You only need to create the Firebase instance with the Firebase class and pass it as value prop to the React's Context:
+
+{{< highlight javascript "hl_lines=8 11 13" >}}
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import './index.css';
+import * as serviceWorker from './serviceWorker';
+
+import App from './components/App';
+import Firebase, { FirebaseContext } from './components/Firebase';
 
-{{< highlight javascript "hl_lines=3 4 5" >}}
-import { auth } from './firebase';
+ReactDOM.render(
+  <FirebaseContext.Provider value={new Firebase()}>
+    <App />
+  </FirebaseContext.Provider>,
+  document.getElementById('root'),
+);
 
-// Sign Up
-export const doCreateUserWithEmailAndPassword = (email, password) =>
-  auth.createUserWithEmailAndPassword(email, password);
+serviceWorker.unregister();
 {{< /highlight >}}
 
-Second, the sign in function which takes email and password parameters, as well. It also uses a Firebase endpoint to sign in a user.
+Doing it this way, we can be assured that Firebase is only instantiated once and that it is injected via React's Context API to React's component tree. Now, every component that is interested in using Firebase has access to the Firebase instance with a `FirebaseContext.Consumer` component. Even though you will see it first-hand later for this application, the following code snippet shows how it would work:
 
-{{< highlight javascript "hl_lines=7 8 9" >}}
-import { auth } from './firebase';
+{{< highlight javascript >}}
+import React from 'react';
 
-// Sign Up
-export const doCreateUserWithEmailAndPassword = (email, password) =>
-  auth.createUserWithEmailAndPassword(email, password);
+import  { FirebaseContext } from '../Firebase';
+
+const SomeComponent = () => (
+  <FirebaseContext.Consumer>
+    {firebase => {
+      return <div>I've access to Firebase and render something.</div>;
+    }}
+  </FirebaseContext.Consumer>
+);
 
-// Sign In
-export const doSignInWithEmailAndPassword = (email, password) =>
-  auth.signInWithEmailAndPassword(email, password);
+export default SomeComponent;
 {{< /highlight >}}
 
-At this point, note that these endpoints are called asynchronously. They need to be resolved later on. In addition, there needs to be error handling for it. For instance, it is not possible to sign in a user who is not signed up yet. The Firebase API would return an error. You will implement all of this in a later part of this tutorial.
+Firebase and React are now connected,  the fundamental step to make the layers communicate with one other. Next, we will implement the interface for the Firebase class on our side to communicate with the Firebase API.
```

Comment:
"with one other" <- correct English?

Reply:
Should have been "with each other" or "with one another"